--- a/hotspot/make/linux/Makefile.orig	2017-03-06 20:12:11.000000000 +0100
+++ a/hotspot/make/linux/Makefile	2017-05-08 07:59:49.775242564 +0200
@@ -349,7 +349,7 @@
 # ENABLE_FULL_DEBUG_SYMBOLS value is used.
 docs: checks
 	$(QUIETLY) mkdir -p $(SUBDIR_DOCS)
-	$(MAKE) -f $(GAMMADIR)/make/$(OSNAME)/makefiles/jvmti.make $(MFLAGS) $(BUILDTREE_VARS) JvmtiOutDir=$(SUBDIR_DOCS) BUILD_FLAVOR=product jvmtidocs
+	#$(MAKE) -f $(GAMMADIR)/make/$(OSNAME)/makefiles/jvmti.make $(MFLAGS) $(BUILDTREE_VARS) JvmtiOutDir=$(SUBDIR_DOCS) BUILD_FLAVOR=product jvmtidocs
 
 # Synonyms for win32-like targets.
 compiler2:  debug product
--- a/hotspot/make/linux/makefiles/defs.make.orig	2017-03-06 20:12:11.000000000 +0100
+++ a/hotspot/make/linux/makefiles/defs.make	2017-05-08 11:28:50.001500588 +0200
@@ -234,7 +234,7 @@
 # Library suffix
 LIBRARY_SUFFIX=so
 
-EXPORT_LIST += $(EXPORT_DOCS_DIR)/platform/jvmti/jvmti.html
+#EXPORT_LIST += $(EXPORT_DOCS_DIR)/platform/jvmti/jvmti.html
 
 # client and server subdirectories have symbolic links to ../libjsig.so
 EXPORT_LIST += $(EXPORT_JRE_LIB_ARCH_DIR)/libjsig.$(LIBRARY_SUFFIX)
